Enhancement of Grocery List with Offline Support and Data Synchronization

DanielRendox opened the pull request:
## Brief description
This PR introduces several key enhancements and fixes to the application. New features include support for grocery item images, lifecycle-aware state collection in UI components, and sorting of grocery items by category and alphabetically within categories. The UI has been improved with animated visibility for the toolbar and a new grocery presentation model to decouple UI from data models. The app now supports offline functionality with local JSON data sources for categories, products, and icons, and includes data synchronization logic with version tracking. A bitmap loading utility has been added for image handling, and DataStore is used for persistent storage of synchronization versions.

## Introduction of Predefined Data

A key part of this PR is the introduction of **predefined data from the assets folder**, which is used to prepopulate the app storage on startup. This data includes grocery categories, default products, and their corresponding images. The purpose of this predefined data is to provide search suggestions when a user types the name of a grocery item in the search bar, enhancing the overall user experience. Some additional assets are already included in the assets folder.

## Handling Data Fetching

The app is designed to periodically connect to the backend when the device is online. However, since the backend for it is not ready yet, the app fetches this data from the assets folder and JSON files. The JSON files are decoded using the **Moshi** library, introduced in this PR, in network classes such as `OfflineGroceryListNetworkDataSource`. This structure allows for easy swapping with another implementation that fetches data from the backend.

## How does the app update its database with new assets?

In addition to initial prepopulation, the app also updates the database when new assets are added. This is achieved through a WorkManager task that runs on each app startup and synchronizes the repositories in the correct order. If any changes are detected, the specified items are updated or deleted. The synchronization logic is unified with the help of helpers in `SyncUtilities`.

## How does the app detect changes in the default data?

Change detection is facilitated by storing versions of icons, products, and categories in Jetpack **Preferences Data Store**, introduced in this pr. If changes are made to the default data in the assets folder, the corresponding change list versions file should also be updated. This allows the app to detect a version difference and fetch the change lists to identify the exact additions, updates, or deletions.

## What is the approach regarding the handling of image data?

It's important to note that the data about icons that flows through the layers of the architecture is only a reference and not an actual image bitmap. The image bitmap is decoded from the corresponding file and displayed to the user only in the view model. This approach may have performance implications due to increased IO operations, which should be measured and addressed if necessary. However, it's not a priority at the moment.

## Why the change in storage from int IDs to string UUIDs?

Lastly, this PR changes the storage of instance IDs to string UUIDs. This is to cater to potential inconsistencies between IDs on the remote and client side, as the app is designed to work with a backend in the future but can also function fully offline. Generating UUIDs on the client side is a simple strategy for ensuring consistency between them, as the chance of generating a duplicate UUID is extremely low.

DanielRendox committed:
Implement update grocery description functionality

DanielRendox committed:
Remove unused getGroceryDescriptions in data layer

This app is not going to support description suggestions anytime soon

DanielRendox committed:
Replace collectAsState with collectAsStateWithLifecycle

For lifecycle-aware flow collection

DanielRendox committed:
Improve toolbar animated visibility

And remove sample icons from the toolbar

DanielRendox committed:
Make not purchased grocery list items sorted by category

Items that are in the same category are sorted in alphabetic order (case insensitive).

DanielRendox committed:
Improve code for more convenient grocery category updates

DanielRendox committed:
Save and retrieve grocery images

DanielRendox committed:
Add grocery presentation with grocery icon bitmaps

to not depend on android-specific components in model and business logic

DanielRendox committed:
Configure Moshi and Retrieve categories from JSON

DanielRendox committed:
Configure retrieving icons from json

Icon references are stored in json files and associated with products through ids. On app startup the internal file storage gets prepopulated with the icons in the assets folder. The room database gets prepopulated with default products and file paths to these icons.

DanielRendox committed:
Fix null icons for search results

DanielRendox committed:
Refactor to store only relative file paths to icons

DanielRendox committed:
Implement updating local data when new version is detected

Checks for new versions, prepopulation, and updates are executed on each app startup.

DanielRendox committed:
Refactor to store ids as Strings instead of Integers

In order to store UUIDs for consistency and easy synchronization across devices.

DanielRendox committed:
Remove log in AssetToFileSaver

DanielRendox commented:
@CodiumAI-Agent /review

CodiumAI-Agent commented:
## PR Review

<table>
<tr>
<tr><td> ⏱️&nbsp;<strong>Estimated&nbsp;effort&nbsp;to&nbsp;review [1-5]</strong></td><td>

5, because the PR introduces a significant amount of changes across various files, including the addition of new features, refactoring of existing code, and updates to the database schema. The complexity and breadth of changes require a thorough review to ensure quality and maintainability.


</td></tr>
<tr><td> 🧪&nbsp;<strong>Relevant tests</strong></td><td>

No


</td></tr>
<tr><td rowspan=3> 🔍&nbsp;<strong>Possible issues</strong></td>
<td>

<strong>Possible Bug:</strong> The absence of tests for the new features and changes introduced in this PR could lead to potential bugs and regressions not being caught early.</td></tr>
<tr>
<td>

<strong>Performance Concern:</strong> The addition of new database entities and the changes in data flow could impact the app's performance. Without proper testing, it's hard to assess the impact.</td></tr>
<tr>
<td>

<strong>Data Migration:</strong> Changes in the database schema, such as the addition of new entities and modifications to existing ones, require careful handling of data migration to prevent data loss or corruption.</td></tr>
<tr><td> 🔒&nbsp;<strong>Security concerns</strong></td><td>

No

</td></tr>
</table>


<details><summary> <strong>Code feedback:</strong></summary>

<hr><table><tr><td>relevant file</td><td>app/src/main/java/com/rendox/grocerygenius/datastore/ChangeListVersionsDataSource.kt
</td></tr><tr><td>suggestion &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</td><td>

<strong>

Consider adding error handling for data store operations to gracefully handle potential issues during data read/write operations. This can improve the robustness of the data layer. [important]

</strong>
</td></tr><tr><td>relevant line</td><td><a href='https://github.com/DanielRendox/GroceryGenius/pull/4/files#diff-485903260a02cdd6cca76a1a0af3a8cd02bd8f42eff66587c2e2edef2aa896b7R15'>"+    suspend fun getChangeListVersions() = userPreferences.data"</a></td></tr></table><hr>

<table><tr><td>relevant file</td><td>app/src/main/java/com/rendox/grocerygenius/network/icons/OfflineIconNetworkDataSource.kt
</td></tr><tr><td>suggestion &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</td><td>

<strong>

Implement caching for icons downloaded from assets to reduce I/O operations and improve UI responsiveness, especially when displaying icons in lists or grids. [important]

</strong>
</td></tr><tr><td>relevant line</td><td><a href='https://github.com/DanielRendox/GroceryGenius/pull/4/files#diff-570357ee3b7a1ad3a42e9d34d2764f97e879411b2bbd29ef946ecb6bcd26e071R21'>"+    override suspend fun downloadIcons(): List<Icon> ="</a></td></tr></table><hr>

<table><tr><td>relevant file</td><td>app/src/main/java/com/rendox/grocerygenius/data/grocery/GroceryRepositoryImpl.kt
</td></tr><tr><td>suggestion &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</td><td>

<strong>

For methods that interact with the database, consider implementing transaction management to ensure data integrity, especially for operations that involve multiple steps or updates. [medium]

</strong>
</td></tr><tr><td>relevant line</td><td>"+    override suspend fun insertProductAndGrocery("</td></tr></table><hr>

<table><tr><td>relevant file</td><td>app/src/main/java/com/rendox/grocerygenius/ui/components/grocery_list/LazyGroceryGrid.kt
</td></tr><tr><td>suggestion &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</td><td>

<strong>

Optimize the rendering of the grocery grid by using lazy loading techniques and efficient diffing to only update items that have changed. This can significantly improve scroll performance in lists with many items. [medium]

</strong>
</td></tr><tr><td>relevant line</td><td>"+fun LazyGroceryGrid("</td></tr></table><hr>

</details><hr>

<details> <summary><strong>✨ Review tool usage guide:</strong></summary><hr> 

**Overview:**
The `review` tool scans the PR code changes, and generates a PR review which includes several types of feedbacks, such as possible PR issues, security threats and relevant test in the PR. More feedbacks can be [added](https://pr-agent-docs.codium.ai/tools/review/#general-configurations) by configuring the tool.

The tool can be triggered [automatically](https://pr-agent-docs.codium.ai/usage-guide/automations_and_usage/#github-app-automatic-tools-when-a-new-pr-is-opened) every time a new PR is opened, or can be invoked manually by commenting on any PR.
- When commenting, to edit [configurations](https://github.com/Codium-ai/pr-agent/blob/main/pr_agent/settings/configuration.toml#L23) related to the review tool (`pr_reviewer` section), use the following template:
```
/review --pr_reviewer.some_config1=... --pr_reviewer.some_config2=...
```
- With a [configuration file](https://pr-agent-docs.codium.ai/usage-guide/configuration_options/), use the following template:
```
[pr_reviewer]
some_config1=...
some_config2=...
```
    

See the review [usage page](https://pr-agent-docs.codium.ai/tools/review/) for a comprehensive guide on using this tool.


</details>

DanielRendox commented:
@CodiumAI-Agent /describe

DanielRendox commented:
@CodiumAI-Agent /update_changelog

CodiumAI-Agent commented:
**Changelog updates:**

## 2024-03-31

### Added
- Support for grocery item images, including saving and retrieving images.
- Lifecycle-aware state collection in UI components.
- Sorting of grocery items by category and alphabetically within categories.
- Animated visibility for toolbar components.
- Grocery presentation model to decouple UI from data models.
- Local JSON data sources for categories, products, and icons, supporting offline functionality.
- Data synchronization logic for categories, products, and icons, with version tracking.
- Bitmap loading utility for image handling.
- DataStore for persistent storage of synchronization versions.

### Changed
- Refactored grocery, category, and product IDs to use strings instead of integers.
- Updated data models to support new features and improve code readability.
- Enhanced UI components for displaying grocery items with images.
- Improved error handling and logging across data sources and repositories.

### Fixed
- Issue with null icons in search results.
- Various minor bugs and performance issues.


>to commit the new content to the CHANGELOG.md file, please type:
>'/update_changelog --pr_update_changelog.push_changelog_changes=true'

DanielRendox commented:
@CodiumAI-Agent /ask "Summarize the general purpose of the PR, improvements that were made, and suggest a better title"

CodiumAI-Agent commented:
Question: Summarize the general purpose of the PR, improvements that were made, and suggest a better title

Answer:
The general purpose of this PR is to improve the user interface and user experience of the grocery list feature in the application. The improvements include changes to the animation, layout, and state management. The PR introduces the use of `collectAsStateWithLifecycle()` instead of `collectAsState()`, which is a more lifecycle-aware way to collect from a Flow in Compose. It also includes changes to the data model, replacing `Grocery` with `GroceryPresentation` to separate the UI representation from the data model. The PR also removes some unused imports and cleans up the code.

A better title for this PR could be "UI/UX Enhancements and Data Model Refactoring for Grocery List Feature". 

Please note that due to the token budget limit, not all added and modified files were processed. The reviewer should manually check these files for a complete review.

<hr>

<details> <summary><strong>✨ Ask tool usage guide:</strong></summary><hr> 

**Overview:**
The `ask` tool answers questions about the PR, based on the PR code changes.
It can be invoked manually by commenting on any PR:
```
/ask "..."
```

Note that the tool does not have "memory" of previous questions, and answers each question independently.        
        

<table>

<tr><td><details> <summary><strong> More PR-Agent commands</strong></summary><hr> 

> To invoke the PR-Agent, add a comment using one of the following commands:  
> - **/review**: Request a review of your Pull Request.   
> - **/describe**: Update the PR title and description based on the contents of the PR.   
> - **/improve [--extended]**: Suggest code improvements. Extended mode provides a higher quality feedback.   
> - **/ask \<QUESTION\>**: Ask a question about the PR.   
> - **/update_changelog**: Update the changelog based on the PR's contents.   
> - **/add_docs** 💎: Generate docstring for new components introduced in the PR.   
> - **/generate_labels** 💎: Generate labels for the PR based on the PR's contents.   
> - **/analyze** 💎: Automatically analyzes the PR, and presents changes walkthrough for each component.   

>See the [tools guide](https://pr-agent-docs.codium.ai/tools/) for more details.
>To list the possible configuration parameters, add a **/config** comment.   
 


</details></td></tr>

</table>

See the [ask usage](https://pr-agent-docs.codium.ai/tools/ask/) page for a comprehensive guide on using this tool.


</details>

DanielRendox commented:
@CodiumAI-Agent /ask "A large part of this PR introduces default data from the assets folder (including images for groceries) that the app storage is prepopulated with on app startup. Can you explain how this works in details?"

CodiumAI-Agent commented:
Question: A large part of this PR introduces default data from the assets folder (including images for groceries) that the app storage is prepopulated with on app startup. Can you explain how this works in details?

Answer:
The PR diff does not provide direct information about how the app storage is prepopulated with default data from the assets folder on app startup. However, it does hint at the use of a `BitmapLoader` and a `JsonAssetDecoder`, which are likely used to load image and JSON data from the assets folder.

The `BitmapLoader` is likely used to load bitmap images from the assets folder. These images could be used as default images for groceries. The `JsonAssetDecoder` is likely used to decode JSON files that contain default data for the app. This data could include default groceries, categories, or other data needed by the app.

The process of prepopulating the app storage on startup would typically involve reading these files from the assets folder, decoding the data, and then inserting this data into the app's database or other storage solution. This process would likely be done in a background thread to avoid blocking the main thread and slowing down the app startup.

However, without the full context of the additional added and modified files mentioned in the PR, it's hard to provide a more detailed explanation. For a complete understanding, you would need to look at the implementation of the classes mentioned and how they are used during the app startup.

<hr>

<details> <summary><strong>✨ Ask tool usage guide:</strong></summary><hr> 

**Overview:**
The `ask` tool answers questions about the PR, based on the PR code changes.
It can be invoked manually by commenting on any PR:
```
/ask "..."
```

Note that the tool does not have "memory" of previous questions, and answers each question independently.        
        

<table>

<tr><td><details> <summary><strong> More PR-Agent commands</strong></summary><hr> 

> To invoke the PR-Agent, add a comment using one of the following commands:  
> - **/review**: Request a review of your Pull Request.   
> - **/describe**: Update the PR title and description based on the contents of the PR.   
> - **/improve [--extended]**: Suggest code improvements. Extended mode provides a higher quality feedback.   
> - **/ask \<QUESTION\>**: Ask a question about the PR.   
> - **/update_changelog**: Update the changelog based on the PR's contents.   
> - **/add_docs** 💎: Generate docstring for new components introduced in the PR.   
> - **/generate_labels** 💎: Generate labels for the PR based on the PR's contents.   
> - **/analyze** 💎: Automatically analyzes the PR, and presents changes walkthrough for each component.   

>See the [tools guide](https://pr-agent-docs.codium.ai/tools/) for more details.
>To list the possible configuration parameters, add a **/config** comment.   
 


</details></td></tr>

</table>

See the [ask usage](https://pr-agent-docs.codium.ai/tools/ask/) page for a comprehensive guide on using this tool.


</details>

DanielRendox commented:
@CodiumAI-Agent /reflect_and_review

CodiumAI-Agent commented:
Questions to better understand the PR:

1) I see that the `collectAsState` function has been replaced with `collectAsStateWithLifecycle`. Can you confirm if this change was made to ensure that the state collection is lifecycle-aware and thus more efficient?
2) The PR includes changes related to storing only relative file paths to icons. Can you explain the benefits of storing relative paths instead of absolute paths?
3) The PR includes a change to store ids as Strings instead of Integers. Is this change made to accommodate UUIDs for better consistency and easy synchronization across devices?
4) Could you elaborate on the logic behind sorting the grocery items first by their purchase status and then by their category's sorting priority?
5) In the `GroceryListScreenViewModel`, the `editGroceryDescriptionFlow` is used to update the grocery description in the repository. Can you explain why this approach was chosen instead of directly updating the repository when the description changes?

 Please respond to the questions above in the following format:


>/answer
>1) ...
>2) ...
>...

DanielRendox committed:
Rename Icon to IconReference

DanielRendox committed:
Change easing in toolbar visibility animation

Use linear easing

DanielRendox committed:
Change logic of updating local change list versions

When the data is prepopulated, update change list versions to the latest ones from the network. However, if the change list is empty, update to 0.

Because it would be good to somehow indicate that the data is up to date. The value of -1 doesn't suggest that. So update to 0 instead.
